Use CubinLinker for CUDA Minor Version Compatibility

gmarkall opened the pull request:
## Description

This switches to using CubinLinker (from PTXCompiler, but CubinLinker uses PTXCompiler internally) for Minor Version Compatibility. This enables support for all Numba features except linking archives with MVC, in support of use cases such as String UDFs (#11319) with MVC.

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

galipremsagar commented on conda/environments/cudf_dev_cuda11.5.yml:
```diff
@@ -7,6 +7,7 @@ channels:
   - rapidsai-nightly
   - dask/label/dev
   - conda-forge
+  - gmarkall
```
You will need to have the channel added into `build.sh` files in the similar way I did in my PR here: https://github.com/rapidsai/cudf/pull/11617/files

gmarkall replied:
Thanks, I've just seen this whilst I was making a change with a similar aim. I will have a look at the way you did it too.

galipremsagar commented:
CI errors related to :https://github.com/dask/dask/issues/9490

brandon-b-miller commented on python/cudf/cudf/core/udf/__init__.py:
```diff
@@ -28,36 +24,32 @@
+    from . import strings_typing  # isort: skip
+    from . import strings_lowering  # isort: skip
+
+    # add an overload of MaskedType.__init__(string_view, bool)
```
```suggestion
```
Removing this comment should fix the isort/black conflict here.

gmarkall commented:
Now that all the required packages are available in the usual channels (Numba 0.56.2 from conda-forge and CubinLinker from rapidsai) I've removed the "DO NOT MERGE" label. Additionally I've added some changes from Brandon that remove the checks on the driver / runtime versions for enabling string UDFs, as the changes in this branch mean that those checks should no longer be necessary.

Assuming the tests all pass this should be good for a review / merge if the review is OK.

galipremsagar commented:
I opened an integration repo PR: https://github.com/rapidsai/integration/pull/535

shwina commented on ci/gpu/build.sh:
```diff
@@ -266,27 +266,14 @@ py.test -n 8 --cache-clear --basetemp="$WORKSPACE/custreamz-cuda-tmp" --junitxml
-    gpuci_logger "Python py.test retest cuDF UDFs"
-    py.test tests/test_udf_masked_ops.py -n 8 --cache-clear
-fi
+# retest cudf with strings_udf present
+py.test -n 8 --cache-clear --basetemp="$WORKSPACE/strings-udf-cuda-tmp" --junitxml="$WORKSPACE/junit-strings-udf.xml" -v --cov-config=.coveragerc --cov=strings_udf --cov-report=xml:"$WORKSPACE/python/strings_udf/strings-udf-coverage.xml" --cov-report term tests
+gpuci_logger "Python py.test retest cuDF UDFs"
+py.test tests/test_udf_masked_ops.py -n 8 --cache-clear
```
Is this right? Don't we need to `cd $WORKSPACE/python/cudf/cudf` between these two lines?

brandon-b-miller replied:
you're right. Looks like I was a little overzealous with my deletions here @gmarkall , sorry about that!

jakirkham commented:
Thanks all! 🎉
